x-pack/plugins/security_solution/public/assistant/get_comments/stream/stream_observable.ts

```diff
@@ -0,0 +1,112 @@
+  setLoading: Dispatch<SetStateAction<boolean>>
+): Observable<PromptObservableState> =>
+  new Observable<PromptObservableState>((observer) => {
+    observer.next({ chunks: [], loading: true });
+    const decoder = new TextDecoder();
+    const chunks: Chunk[] = [];
+    function read() {
+      reader
+        .read()
+        .then(({ done, value }: { done: boolean; value?: Uint8Array }) => {
+          try {
+            if (done) {
+              observer.next({
+                chunks,
+                message: getMessageFromChunks(chunks),
+                loading: false,
+              });
+              observer.complete();
+              return;
+            }
+
+            const nextChunks: Chunk[] = decoder
+              .decode(value)
+              .split('\n')
+              // every line starts with "data: ", we remove it and are left with stringified JSON or the string "[DONE]"
+              .map((str) => str.substring(6))
+              // filter out empty lines and the "[DONE]" string
+              .filter((str) => !!str && str !== '[DONE]')
+              .map((line) => JSON.parse(line));
+
+            nextChunks.forEach((chunk) => {
+              chunks.push(chunk);
+              observer.next({
+                chunks,
+                message: getMessageFromChunks(chunks),
+                loading: true,
+              });
+            });
+          } catch (err) {
+            observer.error(err);
+            return;
+          }
+          read();
+        })
+        .catch((err) => {
+          observer.error(err);
+        });
+    }
+    read();
+    return () => {
+      reader.cancel();
+    };
+  }).pipe(
+    // make sure the request is only triggered once,
+    // even with multiple subscribers
+    shareReplay(1),
+    // append a timestamp of when each value was emitted
+    timestamp(),
+    // use the previous timestamp to calculate a target
+    // timestamp for emitting the next value
+    scan((acc, value) => {
+      const lastTimestamp = acc.timestamp || 0;
+      const emitAt = Math.max(lastTimestamp + MIN_DELAY, value.timestamp);
+      return {
+        timestamp: emitAt,
+        value: value.value,
+      };
+    }),
+    // add the delay based on the elapsed time
+    // using concatMap(of(value).pipe(delay(50))
+    // leads to browser issues because timers
+    // are throttled when the tab is not active
+    concatMap((value) => {
+      const now = Date.now();
+      const delayFor = value.timestamp - now;
+
+      if (delayFor <= 0) {
+        return of(value.value);
+      }
+
+      return of(value.value).pipe(delay(delayFor));
+    }),
+    // set loading to false when the observable completes or errors out
+    finalize(() => setLoading(false))
+  );
+
+function getMessageFromChunks(chunks: Chunk[]) {
+  return chunks.map((chunk) => chunk.choices[0]?.delta.content ?? '').join('');
+}
+
+export const getDumbObservable = () => new Observable<PromptObservableState>();
```

Comment:
nit: clarify const name -- this is just a placeholder or passthrough observable for when `content == null && reader != null`?